form/form.go

```diff
@@ -35,12 +38,63 @@ type Filler interface {
 	Fill(f Form) (map[string]interface{}, error)
 }
 
-// PromptingFiller fills out a set of environschema fields by prompting
-// the user for each field in sequence.
-type PromptingFiller struct {
-	// The Prompter to use to get the responses. If this is nil then
-	// DefaultPrompter will be used.
-	Prompter Prompter
+// SortedFields returns the given fields sorted first by group name.
+// Those in the same group are sorted so that secret fields come after
+// non-secret ones, finally the fields are sorted by description.
+func SortedFields(fields environschema.Fields) []NamedAttr {
+	fs := make(namedAttrSlice, 0, len(fields))
+	for k, v := range fields {
+		fs = append(fs, NamedAttr{
+			Name: k,
+			Attr: v,
+		})
+	}
+	sort.Sort(fs)
+	return fs
+}
+
+// NamedAttr associates a name with an environschema.Field.
+type NamedAttr struct {
+	Name string
+	environschema.Attr
+}
+
+type namedAttrSlice []NamedAttr
+
+func (s namedAttrSlice) Len() int {
+	return len(s)
+}
+
+func (s namedAttrSlice) Less(i, j int) bool {
+	a1 := &s[i]
+	a2 := &s[j]
+	if a1.Group != a2.Group {
+		return a1.Group < a2.Group
+	}
+	if a1.Secret != a2.Secret {
+		return a2.Secret
+	}
+	return a1.Description < a2.Description
+}
+
+func (s namedAttrSlice) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// DefaultPrompter is the default Prompter used by a PromptingFiller when
+// Prompter has not been set.
+var DefaultFiller Filler = IOFiller{
+	In:  os.Stdin,
+	Out: os.Stderr,
+}
+
+// IOPrompter is a Prompter based around an io.Reader and io.Writer.
+type IOFiller struct {
+	// In is used to read responses from the user.
+	In io.Reader
```

Comment:
As a thought we could (instead of having the default above) make the zero value IO filler work by using os.Stdin and os.Stdout by default if they are not configured to anything else. (I like it when zero values work)